File: turtle-project/TurtleGraphicsFSharp/PartOne.fs
namespace TurtleGraphics


[<AutoOpen>]
module PartOne =
    type Value = float

    type Cmd =
        | Forward of Value
        | Left    of Value
        | Right   of Value  

    type Program = list<Cmd>

    type Vec2 = float * float
    type Angle = float

    type TurtleState = 
        {
            direction : Angle     // direction in degrees
            position : Vec2       // current position
            trail    : list<Vec2> // points produced so far
        }

    let interpretTurtleProgram (s : TurtleState) (commands : Program) =
        let runCommand s cmd =
            match cmd with
                | Forward d -> 
                    let (px,py) = s.position
                    let angleInRadians = s.direction * ((2.0 * System.Math.PI) / 360.0)
                    let newPos = px + d * cos angleInRadians, py + d * sin angleInRadians
                    { s with position = newPos; trail = newPos :: s.trail }
                | Right angle ->
                    { s with direction = s.direction + angle }
                | Left angle -> 
                    { s with direction = s.direction - angle }
        List.fold runCommand s commands


    module Examples =
    
        let quad = 
            let program =
                [ Forward 30.0; Left 90.0; Forward 30.0; Left 90.0; 
                  Forward 30.0; Left 90.0; Forward 30.0 ]
            program, (50.0,50.0)

        let spiral =
            let program =
                [
                    for lineLen in [100.0 .. (-2.00) .. 0.0] do
                        yield Forward lineLen
                        yield Left 90.0
                ]
            program, (0.0,0.0)

    let runTurtleProgram startPos (p : Program) : list<Vec2> =
        let initialState = { direction = 90.0; position = startPos; trail = [startPos] }
        let resultState = interpretTurtleProgram initialState p
        resultState.trail |> List.rev


    module Parser =

        open FParsec

        let pForward : Parser<Cmd,unit> = 
            skipString "Forward(" >>. pfloat .>> pchar ')' |>> Forward
        
        let pRight : Parser<Cmd,unit> = 
            skipString "Right(" >>. pfloat .>> pchar ')' |>> Right

        let pLeft : Parser<Cmd,unit> = 
            skipString "Left(" >>. pfloat .>> pchar ')' |>> Left

        let pCmd : Parser<Cmd,unit> =
            pForward <|> pRight <|> pLeft

        let pProgram : Parser<Program,unit> =
            let cmdWithSemicolon = 
                pCmd .>> spaces .>> pchar ';' .>> spaces
            manyTill cmdWithSemicolon eof

        let parseProgram (s : string) =
            run pProgram s 


        module Examples = 
        
            let test1 = """Forward(30.0); Left(90.0); Forward(30.0); Left(90.0); 
    Forward(30.0); Left(90.0); f Forward(30.0);"""

File: turtle-project/turtleGraphicsHaskell/app/PartOne.hs
{-# LANGUAGE FlexibleContexts #-}
module PartOne where

import Text.Parsec

---------------------------------------------------------------
-- Domain
---------------------------------------------------------------

type Value = Float

data Cmd = CForward Value | CLeft Value | CRight Value deriving Show
type Program = [Cmd]

type Vec2 = (Float, Float)
type Angle = Float

data TurtleState = TurtleState {
    direction :: Angle,
    position  :: Vec2,
    trail     :: [Vec2]
} deriving Show

---------------------------------------------------------------
-- Interpreter
---------------------------------------------------------------

-- | Convert degree to radiant
rad :: Float -> Float
rad x = x/180*pi

-- | Cos of degree
cos' :: Float -> Float
cos' x = cos $ rad x

-- | Sin of degree
sin' :: Float -> Float
sin' x = sin $ rad x

-- | Calculate coordinate of one point
-- Float    x or y coordinate
-- Float    distance
-- angle    angle
-- (Float -> Float)     function to calculate angle transformation (cos/sin)
calcXorY :: Float -> Float -> Angle -> (Float -> Float) -> Float
calcXorY x d a f = x + d * (f a)

-- Vec2             original pos
-- Float            distance to move
-- Angle            angle to move
calcXYPos :: Vec2 -> Float -> Angle -> Vec2
calcXYPos (x, y) d a = ((calcXorY x d a cos'), (calcXorY y d a sin'))

-- | Helper to run commands for turtle
cmdTurtle' :: TurtleState -> Angle -> Value -> TurtleState
cmdTurtle' ts a v = TurtleState a (calcXYPos (position ts) v a) ((trail ts) ++ [(position ts)])

-- | Turtle single command interpreter
cmdTurtle :: TurtleState -> Cmd -> TurtleState
cmdTurtle ts (CForward v) = cmdTurtle' ts (direction ts) v
cmdTurtle ts (CLeft v)    = cmdTurtle' ts ((direction ts) + v) 0
cmdTurtle ts (CRight v)   = cmdTurtle' ts ((direction ts) - v) 0

-- | Turtle multiple command interpreter
interpretTurtleProgram :: TurtleState -> Program -> TurtleState
interpretTurtleProgram ts []     = ts
interpretTurtleProgram ts (c:cs) = interpretTurtleProgram (cmdTurtle ts c) cs

---------------------------------------------------------------
-- Paser
---------------------------------------------------------------

-- Parse float
-- see: https://www.schoolofhaskell.com/user/stevely/parsing-floats-with-parsec#parsing-floats
number = many1 digit
plus = char '+' *> number
minus = (:) <$> char '-' <*> number
integer = plus <|> minus <|> number

float = fmap rd $ (++) <$> integer <*> decimal
    where rd      = read :: String -> Float
          decimal = option "" $ (:) <$> char '.' <*> number

pValueInParenthesis :: Parsec String () Value
pValueInParenthesis = 
    do 
        char '('
        v <- float
        char ')'
        return v

pSeparator :: Parsec String () ()
pSeparator = 
    do
        spaces
        char ';'
        spaces

-- | Generates parser combinators for CMDs
-- String 			string that input shold have, to be parsed in the command
-- (Value -> Cmd) 	Cmd
generateCmdParser :: String -> (Value -> Cmd) -> Parsec String () Cmd
generateCmdParser s c =
    do
        string s
        v <- pValueInParenthesis
        return $ c v

pForward :: Parsec String () Cmd
pForward = generateCmdParser "Forward" CForward

pLeft :: Parsec String () Cmd
pLeft = generateCmdParser "Left" CLeft

pRight :: Parsec String () Cmd
pRight = generateCmdParser "Right" CRight

pCmd :: Parsec String () Cmd
pCmd = pForward
    <|> pLeft 
    <|> pRight 
    <?> "command 'Forwad', 'Left' or 'Right'"

pProgram :: Parsec String () Program
pProgram = many $ do 
        c <- pCmd 
        eof <|> pSeparator
        return c

parseCmd p = parse p ""

---------------------------------------------------------------
-- Run
---------------------------------------------------------------

runTurtleProgram :: (Float,Float) -> Program -> [Vec2]
runTurtleProgram pos program =
    (trail turtle) ++ [position turtle]
    where turtle = interpretTurtleProgram (TurtleState 90 pos []) program


parseAndRunTurtleProgramm :: (Float,Float) -> String -> [Vec2]
parseAndRunTurtleProgramm pos str = 
    case parseCmd pProgram str of 
        Left a -> error "Parse error"
        Right p -> runTurtleProgram pos p
